### DISCUSSION TITLE
Cross seeding

### DISCUSSION THREAD

**Discussion by ludviglundgren** [autobrr/autobrr#314] [Ideas]
Lets discuss cross-seeding and how it fits into the world of autobrr and the starr apps.

**Comment by quorn23** [comment]
How it fits into autobrr - autobrr will always be faster, than any RSS feed etc, so you would be able to instantly fetch a cross-seed when it shows up. And as it's based on the announcements, it would overall minimize the hits on the tracker apis, which is always a nice side effect.

Now to some conceptual stuff with the arrs. When to trigger it etc.

Currently the two methods used (mostly) are:
- Trigger a search on completion of the torrent (qbit)
- Trigger a search on import in the arrs.

This works, but brings also some issues.
If you trigger it via qbit, the arrs might not be completed with the import yet. This works fine if you don't use the arrs (with categories movement etc).
If you trigger it via the arrs, this works more gracefully, as the arrs are finished with processing and then the search is triggered.

This works fine, besides Seasons. (never tested it with music, might be the same case with Lidarr) Or in other words anything with more than one file (so not a single episode/movie). Here it's the case that the import trigger triggers on every episode of the season (the concept of the import trigger was done for things like trigger a refresh of the media software etc)

The good thing is (specially here, any dev etc please correct me if i'm wrong)

The qbit api allows to check the content. So it's possible to check if there are multiple video files present.
The sonarr api allows to check the grab of the file, which includes the info if it's a season or not. 
_Side note here, the payload of the import trigger sadly doesn't contain that info._

So if both those cases are _true_ autobrr would just need to poll qbit to await the category change on the season (as that's the moment the season is fully imported and done) and start the search then.

If any of those cases are _false_ the search can happen on the import trigger.

**Comment by JohnDoee** [comment]
I can add a way to build a setup in the context of autobrr.

* Run a filter with new stuff you want to auto download and add it to your torrent client
* Run the same filter against multiple other trackers to download .torrent files (potentially trigger autotorrent add)
* On complete, move torrent data and trigger autotorrent scan + add *.torrent

Unless you want to build your own data matching engine here, then outsourcing it to my tool is the best game in town.

What is nice about this way is the simplicity.
For non-new/search data a slew of other tools already exist.

**Reply by zze0s** [reply]
Hi! We recently added `external filters` to filters which allows for running shell scripts and/or post webhooks that can be used as part of the filtering. The idea was to allow for external checks whatever that means, but it also opened up the possibility to run tools like cross-seed as part of filtering.

mmgoodnow/cross-seed now [cross-seed/cross-seed](https://github.com/cross-seed/cross-seed) added a `/api/announce` endpoint where we can send announces and check against it's local db of releases to see if it's a match or not. It only supports one client tho which is a limitation for some. It's currently experimental and it's working ok.

I'm all for more tools and solutions so how would your implementation look? Never used autotorrent so maybe you can clarify some stuff. 
* Is it a cli only or can run as a service? 
* I see it says multi-client support, is that one at a time but different clients or can it support multiple active clients to check releases in etc?

The plan is to have something built into autobrr to at least match with existing releases in connected clients, but since we haven't started to build this we're open to discuss implementations that can allow other tools to work with it.

And yeah, building the data matching engine is a lot of effort to make it good and cover edge cases that can arise. Work was started on a basic one and it works with properly named releases but that's about it. 😄
